include/tudocomp_stat/StatPhase.hpp

```diff
@@ -305,13 +346,17 @@ class StatPhase {
     /// \param key the statistic key or name
     /// \param value the value to log (will be converted to a string)
     template<typename T>
-    inline void log_stat(std::string&& key, const T& value) {
+    inline void log_stat(const std::string& key, const T& value) {
         if (!m_disabled) {
             suppress_memory_tracking guard;
-            m_data->log_stat(std::move(key), value);
+            m_stats[key] = value;
```

Comment:
Hat `m_stats[std::move(key)]` hier Probleme gemacht?

Reply:
Nö, nicht dass ich wüsste. Muss verloren gegangen sein. :-)

Reply:
Bin mir nicht sicher, was mich da geritten hat, meines Wissens hat es keinen besonderen Grund. Ich ändere das mal und schaue, ob es Kompilationsprobleme in tudocomp gibt. Wenn nicht pushe ich es hier mit herein.

Reply:
Ich denke gerade nochmal über den Fall `log_stat("key", json(...))` nach. Das `json` würde ja im Vorfeld erzeugt und daher in der Speichermessung mit auftauchen.

Was würde passieren, wenn ich den Parameter auf `T&& value` ändere? Würde der Compiler die tatsächliche Allokation dadurch "verschieben"? Ich kenne mich da mit den Abläufen nicht so im Detail aus.

Überdies wollte ich `m_stats` noch als `public`-Referenz von außen zugreifbar machen, so dass auch die (m.E. syntaktisch schönere) Variante `phase.stats["key"] = value;` möglich wird.

Reply:
Das folgenden gilt für sowohl für den `key` als auch `value` Parameter:

`T&&` ist nur eine Referenz - die Allokation würde außerhalb der Methode geschehen, gemessen werden, dann in den Methode verschoben werden, und dann nie mehr als Deallokation gelten weil sie hinter dem Guard ist.

Die `T const&` Variante hätte den Vorteil das zwar außerhalb der Methode Allokiert wird, aber dann auch direkt wieder Deallokiert wird, weil die Methode intern hinter dem Guard eine Kopie davon anlegt.

Die Map nach außen hin als Referenz verfügbar zu machen hätte den Nachteil das sowohl die Allokation des Objekts, als auch die der Map (wenn sie zB resized wird) gemessen werden würde, da man kein Guard Objekt darum hat.

---

Ich hätte gerade folgende Idee: Man könnte sowohl eine `T const&` basierte API anbieten für einfache, aber mitgemessenen temporäre Allokationen, und ein guard Objekt um komplizierte Daten in-place vorbereiten zu können:

```c++
{
    StatPhase phase;
    // operation to measure
    // ...
    auto log_guard = phase.log_unmeasured();
    std::stringstream data;
    data << foo << ", " << expensive_op(); 
    log_guard->log("x", data.str());
}
```

Alternative mit closures:

```c++
{
    StatPhase phase;
    // operation to measure
    // ...
    auto log_guard = phase.log_unmeasured_with([&](auto& log_guard) {
        std::stringstream data;
       data << foo << ", " << expensive_op(); 
       log_guard.log("x", data.str());
    });
}
```

Reply:
Sowas ähnliches wäre auch mein Alternativ-Vorschlag gewesen: praktisch einfach ein `suppress_memory_tracking`-Objekt per Funktion abrufbar zu machen, welches dann als Guard fungiert. Man würde `log` dann immer noch über die Phase aufrufen.

Beides ginge vermutlich? Die Closure-Variante macht definitiv auch Sinn.

Reply:
Ich sehe gerade, durch `pause_tracking` und `resume_tracking` ist das sogar schon in unschön möglich. Ich würde sagen, ich baue einen Guard nach außen ein zuzüglich Closure-Variante und setze diese alten Dinger auf deprecated.